source/loader/layers/validation/ur_leak_check.hpp

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+#ifndef UR_LEAK_CHECK_H
+#define UR_LEAK_CHECK_H 1
+
+#include "backtrace.hpp"
+#include "ur_validation_layer.hpp"
+
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#define MAX_BACKTRACE_FRAMES 64
+
+namespace validation_layer {
+struct RefCountContext {
+  private:
+    struct RefRuntimeInfo {
+        int64_t refCount;
+        std::vector<BacktraceLine> backtrace;
+    };
+
+    std::mutex mutex;
+    std::unordered_map<void *, struct RefRuntimeInfo> counts;
+
+    inline struct RefRuntimeInfo createRefInfo(int refCount) {
+        std::vector<BacktraceLine> backtrace = getCurrentBacktrace();
+
+        return RefRuntimeInfo{refCount, backtrace};
+    }
+
+    std::pair<void *, int64_t> incrementRefCount(void *ptr) {
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            it->second.refCount++;
+            return std::make_pair(it->first, it->second.refCount);
+        } else {
+            counts.emplace(ptr, createRefInfo(1));
+            return std::make_pair(ptr, 1);
+        }
+    }
+
+    std::pair<void *, int64_t> decrementRefCount(void *ptr) {
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            it->second.refCount--;
+            return std::make_pair(it->first, it->second.refCount);
+        } else {
+            counts.emplace(ptr, createRefInfo(-1));
+            return std::make_pair(ptr, -1);
+        }
+    }
+
+  public:
+    enum RefCountUpdateType {
+        REFCOUNT_INCREASE,
+        REFCOUNT_DECREASE,
+    };
+
+    void updateRefCount(void *ptr, enum RefCountUpdateType type) {
```

Comment:
nit: these functions should be the other way around:

```
private:
  updateRefcount(ptr, type) {
    lock lock();
    if counts doesn't have ptr {
       counts.insert(ptr);
    }
    int value = 0;
    match(type) {
      case INC:
        value = 1;
       ...
    }
    counts[ptr] += value;
  }
public:
 inc(ptr) {
    update(ptr, INC);
 }
 dec(ptr) {
    update(ptr, DEC);
 }
```

At least, I think that would be cleaner and would reduce the repetitions and abstraction leaks you have here. Also, notice that you created an implicit requirement to acquire the mutex before calling the inc/dec functions. This isn't ideal - you should always try to keep the scope of the locks narrow and as close to the data they are protecting as possible.

Reply:
Understood, I've reversed the dependency of those functions.